From e20e51e04d99cfccccb7ccc3b0951fe8db8f72a8 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Thu, 25 Sep 2025 14:11:58 +0800
Subject: [PATCH 123/128] mtk: mac80211: Add debugfs group key info for each
 link or bss

Add debugfs group key info for each link or bss. For mld bss, mac80211
would have a additional link directory.
e.g. netdev:ap-mld-1/link-0/XX-key.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 net/mac80211/debugfs_netdev.c | 74 +++++++++++++++++++++++++++++++++++
 net/mac80211/debugfs_netdev.h | 10 +++++
 net/mac80211/ieee80211_i.h    |  3 ++
 net/mac80211/key.c            |  2 +
 4 files changed, 89 insertions(+)

diff --git a/net/mac80211/debugfs_netdev.c b/net/mac80211/debugfs_netdev.c
index e8d4452a..b353db7b 100644
--- a/net/mac80211/debugfs_netdev.c
+++ b/net/mac80211/debugfs_netdev.c
@@ -1038,6 +1038,7 @@ void ieee80211_debugfs_remove_netdev(struct ieee80211_sub_if_data *sdata)
 	if (!sdata->vif.debugfs_dir)
 		return;
 
+	ieee80211_debugfs_group_key_remove(&sdata->deflink, -1);
 	debugfs_remove_recursive(sdata->vif.debugfs_dir);
 	sdata->vif.debugfs_dir = NULL;
 	sdata->debugfs.subdir_stations = NULL;
@@ -1096,6 +1097,7 @@ void ieee80211_link_debugfs_remove(struct ieee80211_link_data *link)
 		return;
 	}
 
+	ieee80211_debugfs_group_key_remove(link, -1);
 	debugfs_remove_recursive(link->debugfs_dir);
 	link->debugfs_dir = NULL;
 }
@@ -1119,8 +1121,80 @@ void ieee80211_link_debugfs_drv_remove(struct ieee80211_link_data *link)
 		return;
 
 	/* Recreate the directory excluding the driver data */
+	ieee80211_debugfs_group_key_remove(link, -1);
 	debugfs_remove_recursive(link->debugfs_dir);
 	link->debugfs_dir = NULL;
 
 	ieee80211_link_debugfs_add(link);
 }
+
+void ieee80211_debugfs_group_key(struct ieee80211_link_data *link,
+				 struct ieee80211_key *key)
+{
+	char buf[32];
+
+	if (!link || !link->debugfs_dir)
+		return;
+
+	ieee80211_debugfs_group_key_remove(link, key->conf.keyidx);
+	snprintf(buf, sizeof(buf), "%s../keys/%d",
+		 link->sdata->vif.valid_links ? "../" : "", key->debugfs.cnt);
+
+	switch (key->conf.keyidx)
+	{
+	case 1:
+	case 2:
+	case 3:
+		link->multicast_key = debugfs_create_symlink("multicast_key",
+							     link->debugfs_dir, buf);
+		break;
+	case 4:
+	case 5:
+		link->mgmt_key = debugfs_create_symlink("mgmt_key",
+							link->debugfs_dir, buf);
+		break;
+	case 6:
+	case 7:
+		link->beacon_key = debugfs_create_symlink("beacon_key",
+							  link->debugfs_dir, buf);
+		break;
+	default:
+		return;
+	}
+}
+
+void ieee80211_debugfs_group_key_remove(struct ieee80211_link_data *link, s8 keyidx)
+{
+	if (!link || !link->debugfs_dir)
+		return;
+
+	switch (keyidx)
+	{
+	case -1:
+		debugfs_remove(link->multicast_key);
+		link->multicast_key = NULL;
+		debugfs_remove(link->mgmt_key);
+		link->mgmt_key = NULL;
+		debugfs_remove(link->beacon_key);
+		link->beacon_key = NULL;
+		break;
+	case 1:
+	case 2:
+	case 3:
+		debugfs_remove(link->multicast_key);
+		link->multicast_key = NULL;
+		break;
+	case 4:
+	case 5:
+		debugfs_remove(link->mgmt_key);
+		link->mgmt_key = NULL;
+		break;
+	case 6:
+	case 7:
+		debugfs_remove(link->beacon_key);
+		link->beacon_key = NULL;
+		break;
+	default:
+		return;
+	}
+}
\ No newline at end of file
diff --git a/net/mac80211/debugfs_netdev.h b/net/mac80211/debugfs_netdev.h
index 20ad7600..caf6c34b 100644
--- a/net/mac80211/debugfs_netdev.h
+++ b/net/mac80211/debugfs_netdev.h
@@ -21,6 +21,10 @@ void ieee80211_link_debugfs_remove(struct ieee80211_link_data *link);
 
 void ieee80211_link_debugfs_drv_add(struct ieee80211_link_data *link);
 void ieee80211_link_debugfs_drv_remove(struct ieee80211_link_data *link);
+void ieee80211_debugfs_group_key(struct ieee80211_link_data *link,
+				 struct ieee80211_key *key);
+void ieee80211_debugfs_group_key_remove(struct ieee80211_link_data *link,
+					s8 keyidx);
 #else
 static inline void ieee80211_debugfs_remove_netdev(
 	struct ieee80211_sub_if_data *sdata)
@@ -40,6 +44,12 @@ static inline void ieee80211_link_debugfs_drv_add(struct ieee80211_link_data *li
 {}
 static inline void ieee80211_link_debugfs_drv_remove(struct ieee80211_link_data *link)
 {}
+static inline void ieee80211_debugfs_group_key(struct ieee80211_link_data *link,
+					       struct ieee80211_key *key)
+{}
+static inline void ieee80211_debugfs_group_key_remove(struct ieee80211_link_data *link,
+						      s8 keyidx)
+{}
 #endif
 
 #endif /* __IEEE80211_DEBUGFS_NETDEV_H */
diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index 07022c07..f851523f 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -1117,6 +1117,9 @@ struct ieee80211_link_data {
 
 #ifdef CPTCFG_MAC80211_DEBUGFS
 	struct dentry *debugfs_dir;
+	struct dentry *multicast_key;
+	struct dentry *mgmt_key;
+	struct dentry *beacon_key;
 #endif
 };
 
diff --git a/net/mac80211/key.c b/net/mac80211/key.c
index ab1e68ea..1842dd14 100644
--- a/net/mac80211/key.c
+++ b/net/mac80211/key.c
@@ -22,6 +22,7 @@
 #include "ieee80211_i.h"
 #include "driver-ops.h"
 #include "debugfs_key.h"
+#include "debugfs_netdev.h"
 #include "aes_ccm.h"
 #include "aes_cmac.h"
 #include "aes_gmac.h"
@@ -946,6 +947,7 @@ int ieee80211_key_link(struct ieee80211_key *key,
 	if (!ret) {
 		ieee80211_debugfs_key_add(key);
 		ieee80211_key_destroy(old_key, delay_tailroom);
+		ieee80211_debugfs_group_key(link, key);
 	} else {
 		ieee80211_key_free(key, delay_tailroom);
 	}
-- 
2.45.2

